e2e/terraform/provision-infra/hcp_vault.tf

```diff
@@ -11,42 +11,87 @@ data "hcp_vault_cluster" "e2e_shared_vault" {
-  policy = templatefile("${path.module}/provision-nomad/etc/acls/vault/nomad-policy.hcl", {
-    role = "nomad-tasks-${local.random_name}"
-  })
+// Use stable naming formatting, so that e2e tests can rely on the
+// CLUSTER_UNIQUE_IDENTIFIER env var to re-build these names when they need to.
+//
+// If these change, downstream tests will need to be updated as well, most
+// notably vaultsecrets.
+locals {
+  workload_identity_path   = "jwt-nomad-${local.random_name}"
+  workload_identity_role   = "jwt-nomad-${local.random_name}-workloads"
+  workload_identity_policy = "jwt-nomad-${local.random_name}-workloads"
 }
 
-resource "vault_token" "nomad" {
-  policies  = [vault_policy.nomad.name]
-  no_parent = true
-  renewable = true
-  ttl       = "72h"
+// The authentication backed is used by Nomad to generated workload identities
+// for allocations.
+//
+// Nomad is running TLS, so we must pass the CA and HTTPS endpoint. Due to
+// limitations within Vault at the moment, the Nomad TLS configuration must set
+// "verify_https_client=false". Vault will return an error without this when
+// writing the auth backend.
+resource "vault_jwt_auth_backend" "nomad_cluster" {
+  depends_on         = [null_resource.bootstrap_nomad_acls]
+  default_role       = local.workload_identity_role
+  jwks_url           = "https://${aws_instance.server[0].private_ip}:4646/.well-known/jwks.json"
+  jwks_ca_pem        = tls_self_signed_cert.ca.cert_pem
+  jwt_supported_algs = ["RS256"]
+  path               = local.workload_identity_path
 }
 
-# The default role that Nomad will use for derived tokens. It's not allowed
-# access to nomad-policy so that it can only mint tokens for tasks, not for new
-# clusters
-resource "vault_token_auth_backend_role" "nomad_cluster" {
-  role_name           = "nomad-tasks-${local.random_name}"
-  disallowed_policies = [vault_policy.nomad.name]
-  orphan              = true
-  token_period        = "259200"
-  renewable           = true
-  token_max_ttl       = "0"
+// This is our default role for the nomad JWT authentication backend within
+// Vault.
+resource "vault_jwt_auth_backend_role" "nomad_cluster" {
+  backend                 = vault_jwt_auth_backend.nomad_cluster.path
+  bound_audiences         = ["vault.io"]
+  role_name               = local.workload_identity_role
+  role_type               = "jwt"
+  token_period            = 1800
+  token_policies          = [local.workload_identity_policy]
+  token_type              = "service"
+  user_claim              = "/nomad_job_id"
+  user_claim_json_pointer = true
+
+  claim_mappings = {
+    nomad_namespace = "nomad_namespace"
+    nomad_job_id    = "nomad_job_id"
+    nomad_task      = "nomad_task"
+  }
+}
+
+// We must use a data source to read the auth backend accessor which is then
+// interpolated into the policy document.
+data "vault_auth_backend" "nomad_cluster" {
+  depends_on = [vault_jwt_auth_backend.nomad_cluster]
+  path       = local.workload_identity_path
+}
```

Comment:
I'm pretty sure this attribute is already exported from `vaut_jwt_auth_backend.nomad_cluster` (ref https://registry.terraform.io/providers/hashicorp/vault/latest/docs/resources/jwt_auth_backend.html#attributes-reference). So we can save ourselves an API call:

```suggestion
```

Reply:
Oh nice, updated within Removed the comment within https://github.com/hashicorp/nomad/pull/25139/commits/c3ef5f1b98720cc24ebd5e8edb494a8c4ecd54e4